File: framework/source/nndeploy/device/device.cc
#include "nndeploy/device/device.h"

namespace nndeploy {
namespace device {

Architecture::Architecture(base::DeviceTypeCode device_type_code)
    : device_type_code_(device_type_code) {};

Architecture::~Architecture() {
  for (auto iter : devices_) {
    if (iter.second != nullptr) {
      delete iter.second;
    }
  }
  devices_.clear();
};

base::Status Architecture::disableDevice() {
  std::lock_guard<std::mutex> lock(mutex_);
  for (auto iter : devices_) {
    if (iter.second != nullptr) {
      iter.second->deinit();
    }
  }
  return base::kStatusCodeOk;
}

base::DeviceTypeCode Architecture::getDeviceTypeCode() const {
  return device_type_code_;
}

base::Status Architecture::insertDevice(int device_id, Device *device) {
  std::lock_guard<std::mutex> lock(mutex_);
  devices_[device_id] = device;
  return base::kStatusCodeOk;
}

std::map<base::DeviceTypeCode, std::shared_ptr<Architecture>> &
getArchitectureMap() {
  static std::once_flag once;
  static std::shared_ptr<
      std::map<base::DeviceTypeCode, std::shared_ptr<Architecture>>>
      architecture_map;
  std::call_once(once, []() {
    architecture_map.reset(
        new std::map<base::DeviceTypeCode, std::shared_ptr<Architecture>>);
  });
  return *architecture_map;
}

base::DataFormat Device::getDataFormatByShape(const base::IntVector &shape) {
  auto shape_size = shape.size();
  if (shape_size == 5) {
    return base::DataFormat::kDataFormatNCDHW;
  } else if (shape_size == 4) {
    return base::DataFormat::kDataFormatNCHW;
  } else if (shape_size == 3) {
    return base::DataFormat::kDataFormatNCL;
  } else if (shape_size == 2) {
    return base::DataFormat::kDataFormatNC;
  } else if (shape_size == 1) {
    return base::DataFormat::kDataFormatN;
  }
  return base::DataFormat::kDataFormatNotSupport;
}

void *Device::allocatePinned(size_t size) { return nullptr; }
void *Device::allocatePinned(const BufferDesc &desc) { return nullptr; }

void Device::deallocatePinned(void *ptr) { return; }

void *Device::getContext() { return nullptr; }
base::Status Device::bindThread() { return base::kStatusCodeOk; }

Stream *Device::createStream() { return nullptr; }
Stream *Device::createStream(void *stream) { return nullptr; }
base::Status Device::destroyStream(Stream *stream) {
  return base::kStatusCodeOk;
}

Event *Device::createEvent() { return nullptr; }
base::Status Device::destroyEvent(Event *event) { return base::kStatusCodeOk; }
base::Status Device::createEvents(Event **events, size_t count) {
  return base::kStatusCodeOk;
}
base::Status Device::destroyEvents(Event **events, size_t count) {
  return base::kStatusCodeOk;
}

base::DeviceType Device::getDeviceType() const { return device_type_; }

// Stream
Stream::Stream(Device *device) : device_(device) {}

Stream::Stream(Device *device, void *stream)
    : device_(device), is_external_(true) {}

Stream::~Stream() {}

base::DeviceType Stream::getDeviceType() const {
  return device_->getDeviceType();
}

Device *Stream::getDevice() const { return device_; }

base::Status Stream::synchronize() { return base::kStatusCodeOk; }

base::Status Stream::recordEvent(Event *event) { return base::kStatusCodeOk; }

base::Status Stream::waitEvent(Event *event) { return base::kStatusCodeOk; }

base::Status Stream::onExecutionContextSetup() { return base::kStatusCodeOk; }

base::Status Stream::onExecutionContextTeardown() {
  return base::kStatusCodeOk;
}

void *Stream::getNativeStream() { return nullptr; }

// Event
Event::Event(Device *device) : device_(device) {}

Event::~Event() {}

base::DeviceType Event::getDeviceType() const {
  return device_->getDeviceType();
}

Device *Event::getDevice() const { return device_; }

bool Event::queryDone() { return true; }

base::Status Event::synchronize() { return base::kStatusCodeOk; }

void *Event::getNativeEvent() { return nullptr; }

Architecture *getArchitecture(base::DeviceTypeCode type) {
  auto arch_map = getArchitectureMap();
  auto arch = arch_map.find(type);
  if (arch == arch_map.end()) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n", type);
    return nullptr;
  } else {
    return arch->second.get();
  }
}

std::shared_ptr<Architecture> getArchitectureSharedPtr(
    base::DeviceTypeCode type) {
  auto arch_map = getArchitectureMap();
  auto arch = arch_map.find(type);
  if (arch == arch_map.end()) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n", type);
    return nullptr;
  } else {
    return arch->second;
  }
}

base::DeviceType getDefaultHostDeviceType() {
  base::DeviceType dst(base::kDeviceTypeCodeCpu);

  // #if NNDEPLOY_ARCHITECTURE_X86
  //   dst.code_ = base::kDeviceTypeCodeX86;
  // #elif NNDEPLOY_ARCHITECTURE_ARM
  //   dst.code_ = base::kDeviceTypeCodeArm;
  // #else
  //   dst.code_ = base::kDeviceTypeCodeCpu;
  // #endif

  dst.code_ = base::kDeviceTypeCodeCpu;
  dst.device_id_ = 0;

  return dst;
}

Device *getDefaultHostDevice() {
  base::DeviceType device_type = getDefaultHostDeviceType();
  return getDevice(device_type);
}

bool isHostDeviceType(base::DeviceType device_type) {
  return device_type.code_ == base::kDeviceTypeCodeCpu ||
         device_type.code_ == base::kDeviceTypeCodeX86 ||
         device_type.code_ == base::kDeviceTypeCodeArm;
}

base::Status checkDevice(base::DeviceType device_type,
                         std::string library_path) {
  Architecture *architecture = getArchitecture(device_type.code_);
  if (architecture == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return base::kStatusCodeErrorInvalidValue;
  }
  return architecture->checkDevice(device_type.device_id_, library_path);
}

base::Status enableDevice(base::DeviceType device_type,
                          std::string library_path) {
  Architecture *architecture = getArchitecture(device_type.code_);
  if (architecture == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return base::kStatusCodeErrorInvalidValue;
  }
  return architecture->enableDevice(device_type.device_id_, library_path);
}

Device *getDevice(base::DeviceType device_type) {
  Architecture *architecture = getArchitecture(device_type.code_);
  if (architecture == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return nullptr;
  }
  return architecture->getDevice(device_type.device_id_);
}

Stream *createStream(base::DeviceType device_type) {
  Device *device = getDevice(device_type);
  if (device == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return nullptr;
  }
  return device->createStream();
}
Stream *createStream(base::DeviceType device_type, void *stream) {
  Device *device = getDevice(device_type);
  if (device == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return nullptr;
  }
  return device->createStream(stream);
}

base::Status destroyStream(Stream *stream) {
  if (stream == nullptr) {
    NNDEPLOY_LOGE("Stream is nullptr\n");
    return base::kStatusCodeErrorInvalidValue;
  }
  Device *device = stream->getDevice();
  if (device == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  stream->getDeviceType().code_);
    return base::kStatusCodeErrorInvalidValue;
  }
  return device->destroyStream(stream);
}

Event *createEvent(base::DeviceType device_type) {
  Device *device = getDevice(device_type);
  if (device == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return nullptr;
  }
  return device->createEvent();
}

base::Status destroyEvent(Event *event) {
  if (event == nullptr) {
    NNDEPLOY_LOGE("Event is nullptr\n");
    return base::kStatusCodeErrorInvalidValue;
  }
  Device *device = event->getDevice();
  if (device == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  event->getDeviceType().code_);
    return base::kStatusCodeErrorInvalidValue;
  }
  return device->destroyEvent(event);
}

base::Status createEvents(base::DeviceType device_type, Event **events,
                          size_t count) {
  Device *device = getDevice(device_type);
  if (device == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return base::kStatusCodeErrorInvalidValue;
  }
  return device->createEvents(events, count);
}

base::Status destroyEvents(base::DeviceType device_type, Event **events,
                           size_t count) {
  if (events == nullptr || count == 0) {
    NNDEPLOY_LOGE("Events is nullptr or count is 0\n");
    return base::kStatusCodeErrorInvalidValue;
  }
  Device *device = getDevice(device_type);
  if (device == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n",
                  device_type.code_);
    return base::kStatusCodeErrorInvalidValue;
  }
  return device->destroyEvents(events, count);
}

std::vector<DeviceInfo> getDeviceInfo(base::DeviceTypeCode type,
                                      std::string library_path) {
  Architecture *architecture = getArchitecture(type);
  if (architecture == nullptr) {
    NNDEPLOY_LOGE("Architecture is not registered for device type: %d\n", type);
    return std::vector<DeviceInfo>();
  }
  return architecture->getDeviceInfo(library_path);
}

base::Status disableDevice() {
  auto &architecture_map = getArchitectureMap();
  for (auto iter : architecture_map) {
    base::Status status = iter.second->disableDevice();
    if (status != base::kStatusCodeOk) {
      NNDEPLOY_LOGE("disableDevice failed\n");
      return status;
    }
  }
  return base::kStatusCodeOk;
}

base::Status destoryArchitecture() {
  base::Status status = disableDevice();
  if (status != base::kStatusCodeOk) {
    NNDEPLOY_LOGE("disableDevice failed\n");
    return status;
  }
  auto &architecture_map = getArchitectureMap();
  architecture_map.clear();
  return base::kStatusCodeOk;
}

}  // namespace device
}  // namespace nndeploy

File: python/nndeploy/device/device.py

import nndeploy._nndeploy_internal as _C

import nndeploy.base


class DeviceInfo(_C.device.DeviceInfo):
    def __init__(self):
        super().__init__()
    
    @property
    def device_type(self):
        return self.device_type_
    
    @device_type.setter
    def device_type(self, value):
        self.device_type_ = value
    
    @property
    def is_support_fp16(self):
        return self.is_support_fp16_
    
    @is_support_fp16.setter
    def is_support_fp16(self, value):
        self.is_support_fp16_ = value


class Architecture():
    def __init__(self, *args, **kwargs):
        """
        Constructs an Architecture object.

        The constructor can be called in the following ways:
        1. str: Constructs an Architecture from a string. eg: "cpu", "ascendcl", etc.
        2. DeviceTypeCode: Constructs an Architecture from a DeviceTypeCode enum value.
        3. _C.base.DeviceTypeCode: Constructs an Architecture from a DeviceTypeCode enum value.
        """
        if len(args) == 1 and isinstance(args[0], str):
            device_type_code = nndeploy.base.DeviceTypeCode.from_name(args[0])
        elif len(args) == 1 and isinstance(args[0], nndeploy.base.DeviceTypeCode):
            device_type_code = args[0]
        elif len(args) == 1 and isinstance(args[0], _C.base.DeviceTypeCode):
            device_type_code = nndeploy.base.DeviceTypeCode(args[0])
        elif len(args) == 1 and isinstance(args[0], _C.device.Architecture):
            device_type_code = args[0].get_device_type_code()
        else:
            raise ValueError("Invalid arguments for Architecture constructor")
        # shared_ptr
        self._architecture = _C.device.get_architecture(device_type_code)

    def check_device(self, device_id: int = 0, library_path: str = ""):
        return self._architecture.check_device(device_id, library_path)

    def enable_device(self, device_id: int = 0, library_path: str = ""):
        return self._architecture.enable_device(device_id, library_path)

    def disable_device(self):
        return self._architecture.disable_device()

    def get_device(self, device_id: int):
        return self._architecture.get_device(device_id)

    def get_device_info(self, library_path: str = ""):
        return self._architecture.get_device_info(library_path)

    def get_device_type_code(self):
        return self._architecture.get_device_type_code()
    
    def insert_device(self, device_id: int, device: _C.device.Device):
        return self._architecture.insert_device(device_id, device)

    def __str__(self):           
        return self._architecture.__str__()
    
class Device():
    def __init__(self, *args, **kwargs):
        """
        Constructs a Device object.

        The constructor can be called in the following ways:
        1. Device(device_name_and_id): Constructs a Device from a string in the format "device_name:device_id" (e.g., "cuda:0").
        2. Device(device_name, device_id): Constructs a Device from a string device name and an integer device ID.
        3. Device(device_type_code): Constructs a Device from a DeviceTypeCode enum value. The device ID defaults to 0.
        4. Device(device_type_code, device_id): Constructs a Device from a DeviceTypeCode enum value and an integer device ID.
        5. Device(): Constructs a default Device with DeviceTypeCode.cpu and device ID 0.
        """
        if len(args) == 1 and isinstance(args[0], str):
            device_type = nndeploy.base.DeviceType(args[0])
        elif len(args) == 1 and isinstance(args[0], nndeploy.base.DeviceType):
            device_type = args[0]
        elif len(args) == 1 and isinstance(args[0], _C.base.DeviceType):
            device_type = nndeploy.base.DeviceType(args[0])
        elif len(args) == 1 and isinstance(args[0], _C.device.Device):
            device_type = args[0].get_device_type()
        elif len(args) == 2 and isinstance(args[0], str) and isinstance(args[1], int):
            device_type = nndeploy.base.DeviceType(args[0], args[1])
        elif len(args) == 1 and isinstance(args[0], nndeploy.base.DeviceTypeCode):
            device_type = nndeploy.base.DeviceType(args[0])
        elif len(args) == 2 and isinstance(args[0], nndeploy.base.DeviceTypeCode) and isinstance(args[1], int):
            device_type = nndeploy.base.DeviceType(args[0], args[1])
        elif len(args) == 1 and isinstance(args[0], _C.base.DeviceTypeCode):
            device_type = nndeploy.base.DeviceType(args[0])
        elif len(args) == 2 and isinstance(args[0], _C.base.DeviceTypeCode) and isinstance(args[1], int):
            device_type = nndeploy.base.DeviceType(args[0], args[1])
        elif len(args) == 0:
            device_type = nndeploy.base.DeviceType()
        else:
            raise ValueError("Invalid arguments for Device constructor")
        
        library_path = kwargs.get("library_path", "")
        status = _C.device.enable_device(device_type, library_path)
        if status == nndeploy.base.StatusCode.Ok:
            self._device = _C.device.get_device(device_type)
        else:
            raise ValueError("Failed to enable device")
    
    def get_data_format_by_shape(self, shape):
        return self._device.get_data_format_by_shape(shape)
    
    def to_buffer_desc(self, desc, config):
        return self._device.to_buffer_desc(desc, config)
    
    def allocate(self, size_or_desc):
        """
        Allocate memory that must be freed using deallocate()
        Args:
            size_or_desc: int for allocation size, or BufferDesc for buffer descriptor
        Returns:
            Memory pointer that must be freed using deallocate()
        """
        if isinstance(size_or_desc, int):
            return self._device.allocate(size_or_desc)
        else:
            return self._device.allocate(size_or_desc)
    
    def deallocate(self, ptr):
        return self._device.deallocate(ptr)
    
    def allocate_pinned(self, size_or_desc):
        """
        Allocate pinned memory that must be freed using deallocate_pinned()
        Args:
            size_or_desc: int for allocation size, or BufferDesc for buffer descriptor
        Returns:
            Memory pointer that must be freed using deallocate_pinned()
        """
        if isinstance(size_or_desc, int):
            return self._device.allocate_pinned(size_or_desc)
        else:
            return self._device.allocate_pinned(size_or_desc)
    
    def deallocate_pinned(self, ptr):
        return self._device.deallocate_pinned(ptr)
    
    def copy(self, src, dst, size, stream=None):
        return self._device.copy(src, dst, size, stream)
    
    def download(self, src, dst, size, stream=None):
        return self._device.download(src, dst, size, stream)
    
    def upload(self, src, dst, size, stream=None):
        return self._device.upload(src, dst, size, stream)
    
    def copy_buffer(self, src, dst, stream=None):
        return self._device.copy(src, dst, stream)
    
    def download_buffer(self, src, dst, stream=None):
        return self._device.download(src, dst, stream)
    
    def upload_buffer(self, src, dst, stream=None):
        return self._device.upload(src, dst, stream)
    
    def get_context(self):
        return self._device.get_context()
    
    def create_stream(self, stream=None):
        if stream is None:
            return self._device.create_stream()
        else:
            return self._device.create_stream(stream)
    
    def create_event(self):
        return self._device.create_event()
    
    def create_events(self, events):
        return self._device.create_events(events)
    
    def get_device_type(self):
        return self._device.get_device_type()
    
    def init(self):
        return self._device.init()
    
    def deinit(self):
        return self._device.deinit()
    
    def __str__(self):
        return self._device.__str__()
    

class Stream():
    def __init__(self, device_or_type, stream=None):
        if isinstance(device_or_type, Device):
            device_type = device_or_type.get_device_type()
        else:
            device_type = device_or_type
        self._stream = _C.device.create_stream(device_type, stream)

    def get_device_type(self):
        return self._stream.get_device_type()
    
    def get_device(self):
        return self._stream.get_device()
    
    def synchronize(self):
        return self._stream.synchronize()
    
    def record_event(self, event):
        return self._stream.record_event(event)
    
    def wait_event(self, event):
        return self._stream.wait_event(event)
    
    def on_execution_context_setup(self):
        return self._stream.on_execution_context_setup()
    
    def on_execution_context_teardown(self):
        return self._stream.on_execution_context_teardown()
    
    def get_native_stream(self):
        return self._stream.get_native_stream()
    
    def __str__(self):
        return self._stream.__str__()


class Event():
    def __init__(self, device_or_type):
        if isinstance(device_or_type, Device):
            device_type = device_or_type.get_device_type()
        else:
            device_type = device_or_type
        self._event = _C.device.create_event(device_type)

    def get_device_type(self):
        return self._event.get_device_type()
    
    def get_device(self):
        return self._event.get_device()
    
    def query_done(self):
        return self._event.query_done()
    
    def synchronize(self):
        return self._event.synchronize()
    
    def get_native_event(self):
        return self._event.get_native_event()
    
    def __str__(self):
        return self._event.__str__()
    